### PR TITLE
#5 Server: Pre-parse request branch logic

### PR DESCRIPTION
**Pull request by JonathanLorimer**
Working on parsing request. See issue [here](https://github.com/JonathanLorimer/weft/issues/5)

### CODE REVIEW

**Comment by JonathanLorimer** [comment]
@isovector I am having trouble figuring out how to let the type system know which record to parse. I don't think this is known statically.

the generic type that has to be applied is [here](https://github.com/JonathanLorimer/weft/blob/5-Server-pre-parse-request-branch-logic/src/Server.hs#L20-L23)

the function is called [here](https://github.com/JonathanLorimer/weft/blob/5-Server-pre-parse-request-branch-logic/src/Server.hs#L49)

I get this error, which is the error I would expect to get.
```
    • No instance for (Weft.Generics.EmptyQuery.GEmptyQuery
                         (GHC.Generics.Rep (record 'Query)))
        arising from a use of ‘queryParser’
    • In the first argument of ‘parseOnly’, namely ‘queryParser’
      In the expression: parseOnly queryParser query
      In an equation for ‘parseReqBody’:
          parseReqBody (QueryRequest query) = parseOnly queryParser query
```

Once I figure out how to recognize this type then it should be trivial to apply the `record Resolver` from the Reader as mentioned in this [issue](https://github.com/JonathanLorimer/weft/issues/7)

**Comment by isovector** [comment]
I think you want to parameterize `RequestType` by the record types it can contain. Statically we only want to parse it as the top-level query record.

**Comment by JonathanLorimer** [comment]
That makes sense to me. How does the user define different kinds of queries? i.e. for the user type
```
data User ts = User
  { userId         :: Magic ts (Arg "arg" (Maybe String) -> Id)
  , userName       :: Magic ts Name
  , userBestFriend :: Magic ts (Arg "arg" (Maybe String) -> User ts)
  , userFriends    :: Magic ts [User ts]
  } deriving (Generic)
```
how do we know what kind of top level queries the user wants to allow. For example they might want a GQL schema that looks like this
```
type Query {
     getAllUsers: [User!]!
     getUser(id: $id): User
}
```

I don't think we can generically generate those two options, or even know what the user wants.

I bring this up with regards to parsing the query, because normally the query bodies we would support would look like this:

```
query { 
     getUser( id: "1234516" ) {
          userId
      }
}
query { 
     getUsers {
          userId
          name
      }
}
```

and we need to be able to parse the different query types to a) see if they are a valid query and b) match them with the resolver

**Comment by isovector** [comment]
I think we should just expose one type for the query that the user can choose. That thing could be `User` in our testdata, or it could be a more explicit `Query` sort of thing that exposes whatever top-level things they want.

**Comment by JonathanLorimer** [comment]
Would the user put together a record of valid queries? That makes the most sense to me.

**Comment by isovector** [comment]
Yeah; they just give us any `record 'Resolver` and we hook it up. They can choose what that `record` should look like. What do you think?

**Comment by JonathanLorimer** [comment]
That makes sense to me. Would the top level query name in the graphQL request to our server have to match the fieldname in the record?

**Comment by JonathanLorimer** [comment]
I think I get it now. I am going to pull this [issue](https://github.com/JonathanLorimer/weft/issues/7) into this branch, because there is clearly some overlap.

**Comment by JonathanLorimer** [comment]
# What's Left To Do
1. [ ] ~Parse out queryType name from query body~ (not necessary)
2. [x] Parse out record Query from requestBody
3. [x] call resolve with the resolver and the Query
4. [ ] ~force the user to provide a top-level queryType through the type system (might need another TypeState, although maybe we can just hard code a type that is a record of resolvers)~

Hopefully the type inference will be able to detect the record type from the resolver pulled off the queryType

**Comment by isovector** [comment]
I'm not sure the first or fourth point here are necessary. I think they should just fall out from what we've got already :)

File: package.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
-name: gql
+name: weft
```
Comment (isovector) [inline package.yaml]: Let's do this in a separate PR

File: package.yaml [hunk 2]
```diff
@@ -39,6 +39,7 @@ dependencies:
 
 default-extensions:
   - DeriveGeneric
+  - DeriveAnyClass
```
Comment (isovector) [inline package.yaml]: I'd suggest turning this on only in the modules that need it. It can have weird action-at-a-distance otherwise.

File: src/Server.hs [hunk 3]
```diff
@@ -1,51 +1,61 @@
...
+    encodeUtf8 <$> ((^? key "query" . _String) json)
 
 note :: Maybe a -> Either String a
 note Nothing = Left ""
 note (Just x) = Right x
 
 app :: Application
 app req f = do
-    rb <- requestBody req
+    rb <- getRequestBodyChunk req
```
Comment (isovector) [inline src/Server.hs]: nitpick: trailing spaces

File: src/Server.hs [hunk 4]
```diff
@@ -1,51 +1,61 @@
...
 
 app :: Application
 app req f = do
-    rb <- requestBody req
+    rb <- getRequestBodyChunk req    
     let _eitherQuery = do
             textQuery <- note . maybeQuery $ rb
-            parseReqBody textQuery
-    f $ responseLBS status200 [(hContentType, "text/plain")] "type Query { getGod: God! } type God { name: String entity_type: String god_of: String parents: [Entity] consorts: [Entity] children: [God] roman: String generation: Int olympian: Boolean }"
+            parseReqBody $ Data.ByteString.Char8.concat ["{ ", textQuery, " }"]
```
Comment (isovector) [inline src/Server.hs]: This shouldn't be necessary in the final version. I filed #33 for myself.

File: src/Server.hs [hunk 5]
```diff
@@ -1,51 +1,61 @@
...
+    case _eitherQuery of
+        Right q -> do
+            res <- resolve gqlResolver q
+            f $ successResponse res
+        Left e  -> f $ errorResponse $ BL.fromStrict . pack $ e
```
Comment (isovector) [inline src/Server.hs]: I think this `.` would be easier to read as a `$`.

File: src/TestData.hs [hunk 6]
```diff
@@ -29,6 +39,49 @@ data Account ts = Account
 deriving instance AllHave Show (Account ts) => Show (Account ts)
 deriving instance AllHave Eq (Account ts)   => Eq (Account ts)
 
+jonathan :: User 'Data
```
Comment (isovector) [inline src/TestData.hs:43]: Can we move this stuff into a proper test?

File: src/Weft/Generics/QueryParser.hs [hunk 7]
```diff
@@ -145,17 +146,17 @@ parseRawArgValue = choice
       ]
   ]
 
-parseStringValue :: Parser String
+parseStringValue :: AP.Parser String
```
Comment (isovector) [inline src/Weft/Generics/QueryParser.hs]: Why this change?

File: test/ParserSpec.hs [hunk 8]
```diff
@@ -15,6 +15,12 @@ import           Weft.Internal.Types
 import           Weft.Types
 
 
+
+foo :: Either String (Gql GqlQuery a b 'Query)
```
Comment (isovector) [inline test/ParserSpec.hs]: This isn't used anywhere in the test

**Comment by JonathanLorimer** [comment]
# Final To-Do
- [x] Turn TestData into an actual test
- [ ] Use DerivingVia to generate a JSON that omits Nothings

File: src/Server.hs [hunk 9]
```diff
@@ -1,51 +1,61 @@
...
 import Data.Text.Encoding
+import qualified Data.Map as M
 import qualified Data.ByteString.Lazy as BL
+import Control.Monad.Reader
 
-
-
-
-parseReqBody :: Text -> Either String (User 'Query)
+parseReqBody :: (HasEmptyQuery GqlQuery, HasQueryParser GqlQuery)
```
Comment (JonathanLorimer) [inline src/Server.hs]: @isovector eventually we have to let the user pass in their own GqlQuery type, as well as their GqlMutation and GqlSubscription type

Comment (isovector) [reply src/Server.hs]: Try parsing this as a `Gql query` instead of a `GqlQuery`.

Comment (JonathanLorimer) [reply src/Server.hs]: Solved this kinda. Now we have to give the user a way to pass in their Gql type and the accompanying resolver

File: src/Server.hs [hunk 10]
```diff
@@ -1,51 +1,62 @@
...
+    encodeUtf8 <$> ((^? key "query" . _String) json)
 
 note :: Maybe a -> Either String a
 note Nothing = Left ""
 note (Just x) = Right x
 
-app :: Application
-app req f = do
-    rb <- requestBody req
+app :: (ToJSON (q 'Response), HasEmptyQuery q, HasQueryParser q) => Gql q () () 'Resolver -> Application
```
Comment (isovector) [inline src/Server.hs]: `(ToJSON (q 'Response), Wefty q) =>`

File: src/TestData.hs (hunk 6 above)
Comment (isovector) [reply src/TestData.hs:43]: i'd still like to move this out of Testdata. in my ideal world there will be no testdata file!

File: src/TestData.hs [hunk 11]
```diff
@@ -28,6 +40,34 @@ data Account ts = Account
...
+sandy = User { userId = (Id "2"), userName = ( Name "Sandy"), userBestFriend = jonathan, userFriends = []}
+
+getUserResolver :: (Arg "id" Id) -> User 'Query -> IO (User 'Response)
+getUserResolver a q
+    | (getArg a) == (Id "1") = pure $ hydrate jonathan q
+    | (getArg a) == (Id "2") = pure $ hydrate sandy q
+    | otherwise = pure $ hydrate jonathan q
+
+getAllUsersResolver :: User 'Query -> IO ([User 'Response])
+getAllUsersResolver q = pure $ (flip hydrate q) <$> [sandy, jonathan]
```
Comment (isovector) [inline src/TestData.hs:56]: Is hydrate smart enough that this could be just `pure $ hydrate [sandy, jonathan] q`? If not let's file a bug so it can be.

File: src/Weft/Types.hs [hunk 12]
```diff
@@ -41,3 +44,15 @@ type instance GFields c (f :*: g)  = (GFields c f, GFields c g)
...
 type instance GFields c (K1 i a)   = c a
 
+data Gql q m s (ts :: TypeState) = Gql
+  { query        :: Magic ts (q ts)
+  -- , mutation     :: m ts
+  -- , subscription :: s ts
+  }
+  deriving Generic
+
+deriving instance (Show (q 'Query))       => Show (Gql q m s 'Query)
```
Comment (isovector) [inline src/Weft/Types.hs]: `deriving instance AllHave Show (Gql ts)     => Show (Gql ts)` et al

File: test/ServerSpec.hs [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+deriving instance AllHave Show (User ts)     => Show (User ts)
+deriving instance AllHave Eq (User ts)       => Eq (User ts)
+deriving instance AllHave ToJSON (User ts)   => ToJSON (User ts)
+
+getAllUsersTestString :: ByteString
+getAllUsersTestString = "query { \n getAllUsers { \n userId \n userName \n userFriends { \n userId \n userName \n } \n } \n } \n "
+
+getAllUsersTestQuery :: Either String (Gql GqlQuery () () 'Query)
+getAllUsersTestQuery = Right (Gql { query = Just (ANil
+                                             , GqlQuery { getUser = Nothing
```
Comment (isovector) [inline test/ServerSpec.hs]: Nitpick: possible to get this down so it fits in 80 columns? my mind melts as soon as i need wordwrapping

File: test/ServerSpec.hs [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+                                             )
+                              }
+                         )
+
+getUserTestString :: ByteString
+getUserTestString = " query { \n getUser(id: 1) { \n userId \n userName \n userBestFriend { \n userName \n } \n } \n } \n "
+
+getUserTestQuery :: Either String (Gql GqlQuery () () 'Query)
+getUserTestQuery = Right (Gql { query = Just (ANil
+                                             , GqlQuery { getAllUsers = Nothing
```
Comment (isovector) [inline test/ServerSpec.hs]: likewise 80 columns

File: test/ServerSpec.hs [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+                                                                                    , userFriends = Nothing
+                                                                                    }
+                                                                             )
+                                                        }
+                                             )
+                              }
+                         )
+
+
+deriving instance Show (GqlQuery 'Response)
```
Comment (isovector) [inline test/ServerSpec.hs]: Prefer the same `deriving instance AllHave Show (GqlQuery ts) => Show (GqlQuery ts)` stuff

File: test/ServerSpec.hs [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+deriving instance ToJSON (GqlQuery 'Response)
+deriving instance Show (GqlQuery 'Query)
+
+instance Arbitrary (User 'Query) where
+  arbitrary = recordGen
```
Comment (isovector) [inline test/ServerSpec.hs:111]: I learned the other day that if you also add `shrink = genericShrink` here, counterexamples will come out much more nicely :)

File: src/TestData.hs [hunk 17]
```diff
@@ -1,82 +0,0 @@
-{-# LANGUAGE GeneralizedNewtypeDeriving #-}
```
Comment (isovector) [inline src/TestData.hs]: Love the boldness! Pretty sure deleting this will break the test suite today though :)

Comment (JonathanLorimer) [reply src/TestData.hs]: it doesn't break the test-suite, but I realized it will break the server until we fully implement the ReaderT with all the resolvers on board.

Comment (JonathanLorimer) [reply src/TestData.hs]: Fixed
